tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/io/socket/ServerSocketTest.java

```diff
@@ -108,23 +98,29 @@ private void connectClient(int port, int retryInterval, CompileResult compileRes
             final String clientMsg = "This is the first type of message.";
             boolean connected = false;
             for (int retryCount = 0; retryCount < numberOfRetryAttempts; retryCount++) {
-                try (Socket s = new Socket("localhost", port)) {
-                    BufferedReader input = new BufferedReader(new InputStreamReader(s.getInputStream()));
-                    String answer = input.readLine();
-                    Assert.assertEquals(welcomeMsg.trim(), answer, "Didn't get the expected response from server.");
-                    try (OutputStreamWriter out = new OutputStreamWriter(s.getOutputStream(), StandardCharsets.UTF_8)) {
-                        out.write(clientMsg, 0, clientMsg.length());
-                        out.flush();
-                        out.close();
-                        s.close();
-                        connected = true;
-                        break;
+                try {
+                    SocketChannel socketChannel = SocketChannel.open();
+                    socketChannel.connect(new InetSocketAddress("localhost", port));
+                    ByteBuffer buf = ByteBuffer.allocate(512);
+                    int bytesRead = socketChannel.read(buf);
+                    if (bytesRead == -1) {
+                        socketChannel.close();
+                    }
+                    String answer = new String(buf.array(), StandardCharsets.UTF_8).trim();
+                    Assert.assertEquals(answer, welcomeMsg.trim(), "Didn't get the expected response from server.");
+                    buf.clear();
+                    buf.put(clientMsg.getBytes());
+                    buf.flip();
+                    while (buf.hasRemaining()) {
+                        socketChannel.write(buf);
                     }
+                    connected = true;
+                    socketChannel.close();
```

Comment:
Use finally block to close the socketChannel. There is an assert in this try block, so there is the possibility of this never execute.

Reply:
Ack.